app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.kt

```diff
@@ -0,0 +1,189 @@
+
+        animate(firstStartProgress, popOut()) { this.swapProgressWithSuccessAndContinue() }
+                .start()
+    }
+
+    private fun popOut(): (View) -> Animator = {
+        AnimatorSet().apply {
+            val alphaAnimator = createFadeAnimationFor(it, 1f, 0f)
+            val scaleAnimator = createScaleAnimationFor(it, FastOutLinearInInterpolator(), 1f, 0f)
+            playTogether(alphaAnimator, scaleAnimator)
+            duration = shortAnimationDuration.toLong()
+        }
+    }
+
+    private fun swapProgressWithSuccessAndContinue() {
+        firstStartProgress.setImageResource(R.drawable.ic_circle_tick)
+
+        animate(firstStartProgress, popBackIn()) { this.continueToScheduleAfterDelay() }
+                .start()
+    }
+
+    private fun popBackIn(): (View) -> Animator = {
+        AnimatorSet().apply {
+            val alphaAnimator = createFadeAnimationFor(it, 0f, 1f)
+            val scaleAnimator = createScaleAnimationFor(it, BounceInterpolator(), 0f, 1f)
+            playTogether(alphaAnimator, scaleAnimator)
+            duration = shortAnimationDuration.toLong()
+        }
+    }
+
+    @Suppress("SpreadOperator")   // We cannot avoid using the spread operator here, we use varargs APIs
+    private fun createFadeAnimationFor(view: View, vararg values: Float): Animator {
+        val property = Property.of(View::class.java, Float::class.java, "alpha")
+        val animator = ObjectAnimator.ofFloat(view, property, *values)
+        animator.interpolator = LinearInterpolator()
+        return animator
+    }
+
+    @Suppress("SpreadOperator")   // We cannot avoid using the spread operator here, we use varargs APIs
+    private fun createScaleAnimationFor(view: View, interpolator: Interpolator, vararg values: Float): Animator {
+        val scaleX = Property.of(View::class.java, Float::class.java, "scaleX")
+        val scaleY = Property.of(View::class.java, Float::class.java, "scaleY")
+        val animator = AnimatorSet()
+        animator.playTogether(
+                ObjectAnimator.ofFloat(view, scaleX, *values),
+                ObjectAnimator.ofFloat(view, scaleY, *values)
+        )
+        animator.interpolator = interpolator
+        return animator
+    }
+
+    private fun animate(view: View, animationProducer: (View) -> Animator, endAction: () -> Unit): Animator {
+        val animator = animationProducer.invoke(view)
+        animator.addListener(object : AnimationEndListener {
+            override fun onAnimationEnd(animation: Animator) = endAction.invoke()
+        })
+        return animator
+    }
+
+    private fun continueToScheduleAfterDelay() {
+        firstStartProgress.postDelayed({
+            startActivity(continuationIntent)      // We don't use the navigator here, we basically want to restart the whole flow
+            finish()
+        }, DELAY_AFTER_ANIMATIONS_MILLIS)
+    }
+
+    private inner class NetworkConnectedCallback : ConnectivityManager.NetworkCallback() {
```

Comment:
what's the difference between inner class and private top level classes?

Reply:
https://kotlinlang.org/docs/reference/nested-classes.html

tl;dr it's a non-static inner class, because it can access the outer classes' fields (it's a direct translation from Java)